libraries/ms-drivers/src/ltc_afe_impl.c

```diff
@@ -0,0 +1,302 @@
+  prv_wakeup_idle(afe);
+  return spi_exchange(settings->spi_port, (uint8_t *)&config_packet, len, NULL, 0);
+}
+
+static void prv_calc_offsets(LtcAfeStorage *afe) {
+  // Our goal is to populate result arrays as if the ignored inputs don't exist. This requires
+  // converting the actual LTC6811 cell index to some potentially smaller result index.
+  //
+  // Since we access the same register across multiple devices, we can't just keep a counter and
+  // increment it for each new value we get during register access. Instead, we precompute each
+  // input's corresponding result index. Inputs that are ignored will not be copied into the result
+  // array.
+  //
+  // Similarly, we do the opposite mapping for discharge.
+  LtcAfeSettings *settings = &afe->settings;
+  size_t cell_index = 0;
+  size_t aux_index = 0;
+  for (size_t device = 0; device < settings->num_cells; device++) {
+    for (size_t device_cell = 0; device_cell < LTC_AFE_MAX_CELLS_PER_DEVICE; device_cell++) {
+      size_t cell = device * LTC_AFE_MAX_CELLS_PER_DEVICE + device_cell;
+
+      if ((settings->cell_bitset[device] >> device_cell) & 0x1) {
+        // Cell input enabled - store the index that this input should be stored in
+        // when copying to the result array and the opposite for discharge
+        afe->discharge_cell_lookup[cell_index] = cell;
+        afe->cell_result_lookup[cell] = cell_index++;
+      }
+
+      if ((settings->aux_bitset[device] >> device_cell) & 0x1) {
+        // Cell input enabled - store the index that this input should be stored in
+        // when copying to the result array
+        afe->aux_result_lookup[cell] = aux_index++;
+      }
+    }
+  }
+}
+
+StatusCode ltc_afe_impl_init(LtcAfeStorage *afe, const LtcAfeSettings *settings) {
+  if (settings->num_devices > LTC_AFE_MAX_DEVICES ||
+      settings->num_cells > settings->num_devices * LTC_AFE_MAX_CELLS) {
+    // bad no. devices (needs code change)
+    // bad no. of cells (needs verification)
+    return status_code(STATUS_CODE_INVALID_ARGS);
+  }
+  memset(afe, 0, sizeof(*afe));
+  memcpy(&afe->settings, settings, sizeof(afe->settings));
+
+  prv_calc_offsets(afe);
+  crc15_init_table();
+
+  SpiSettings spi_config = {
+    .baudrate = settings->spi_baudrate,  //
+    .mode = SPI_MODE_3,                  //
+    .mosi = settings->mosi,              //
+    .miso = settings->miso,              //
+    .sclk = settings->sclk,              //
+    .cs = settings->cs,
+  };
+  spi_init(settings->spi_port, &spi_config);
+
+  // TODO(SOFT-9): Update GPIO usage to match updated design
+  // Use GPIO1 as analog input, GPIO2-5 as digital output
+  uint8_t gpio_bits = LTC6811_GPIO1_PD_OFF | LTC6811_GPIO2_PD_ON | LTC6811_GPIO3_PD_ON |
+                      LTC6811_GPIO4_PD_ON | LTC6811_GPIO5_PD_ON;
+  return prv_write_config(afe, gpio_bits);
+}
+
+StatusCode ltc_afe_impl_trigger_cell_conv(LtcAfeStorage *afe) {
+  return prv_trigger_adc_conversion(afe);
```

Comment:
wanna just yoink this function? not necessary but makes things just a lil cleaner

Reply:
yoink? We could move the contents within here, and then call this internally in `ltc_afe_impl_trigger_aux_conv` but I think that would make things more confusing on that side.

Reply:
I see what you mean, makes sense just to leave it here then yeah.